go/vt/vttablet/tabletmanager/restore.go

```diff
@@ -84,27 +84,212 @@ func init() {
+// transitions to and from the RESTORE tablet type, and performs the replication
+// action requested by this functions return values.
+func (tm *TabletManager) restoreBackupInnerLocked(
+	ctx context.Context,
+	logger logutil.Logger,
 	allowedBackupEngines []string,
-	mysqlShutdownTimeout time.Duration) error {
+	deleteBeforeRestore bool,
+	dryRun bool,
+	keyspace string,
+	keyspaceType topodatapb.KeyspaceType,
+	mysqlShutdownTimeout time.Duration,
+	restoreToPos replication.Position,
+	restoreToTimestamp time.Time,
+	startTime time.Time,
+	waitForBackupInterval time.Duration,
+) (
+	replicationAction replicationAction,
+	replicationPos replication.Position,
+	overrideTabletType *topodatapb.TabletType,
+	err error,
+) {
+	tablet := tm.Tablet()
+
+	// Perform the restore. Loop until a backup exists, unless we were told to give up immediately.
+	params := mysqlctl.RestoreParams{
+		Cnf:                  tm.Cnf,
+		Mysqld:               tm.MysqlDaemon,
+		Logger:               logger,
+		Concurrency:          restoreConcurrency,
+		HookExtraEnv:         tm.hookExtraEnv(),
+		DeleteBeforeRestore:  deleteBeforeRestore,
+		DbName:               topoproto.TabletDbName(tablet),
+		Keyspace:             keyspace,
+		Shard:                tablet.Shard,
+		StartTime:            startTime,
+		DryRun:               dryRun,
+		Stats:                backupstats.RestoreStats(),
+		MysqlShutdownTimeout: mysqlShutdownTimeout,
+		AllowedBackupEngines: allowedBackupEngines,
+		RestoreToPos:         restoreToPos,
+		RestoreToTimestamp:   restoreToTimestamp,
+	}
+	var backupManifest *mysqlctl.BackupManifest
+	for {
+		backupManifest, err = mysqlctl.Restore(ctx, params)
+		if backupManifest != nil {
+			statsRestorePosition.Set(replication.EncodePosition(backupManifest.Position))
+			statsRestoreBackupTime.Set(backupManifest.BackupTime)
+		}
+		logger.Infof("Restore: got a restore manifest: %v, err=%v, waitForBackupInterval=%v", backupManifest, err, waitForBackupInterval)
+		if waitForBackupInterval == 0 {
+			break
+		}
+		// Break if no errors.
+		if err == nil {
+			break
+		}
+		// Retry a specific set of errors. The rest we handle immediately.
+		if slices.Contains([]error{mysqlctl.ErrNoBackup, mysqlctl.ErrNoCompleteBackup}, err) {
+			break
+		}
+		log.Infof("No backup found. Waiting %v (from -wait-for-backup-interval flag) to check again.", waitForBackupInterval)
+		select {
+		case <-ctx.Done():
+			err = ctx.Err()
+		case <-time.After(waitForBackupInterval):
+		}
+	}
+
+	// Handle error.
+	if err == nil && backupManifest == nil {
+		err = vterrors.New(vtrpcpb.Code_INTERNAL, "no backup manifest")
+	}
+	if err != nil && err != mysqlctl.ErrNoBackup {
+		return replicationAction, replicationPos, overrideTabletType, vterrors.Wrap(err, "can't restore backup")
+	}
+
+	// Choose replication action.
+	switch {
+	case params.DryRun:
+		// Do nothing here.
+		params.Logger.Infof("Dry run. No changes made")
+		replicationAction = replicationActionNone
```

Comment:
previously, in `restoreDataLocked` we performed the replication action directly. now we signal back to `restoreCommonInnerLocked` the action we want to be performed.